lib/babe/babe.go

```diff
@@ -497,6 +497,21 @@ func (b *Service) handleSlot(epoch, slotNum uint64,
 		return err
 	}
 
+	slotNumOfTheBestBlock, err := b.blockState.GetSlotForBlock(parentHeader.Hash())
+	if err != nil {
+		return fmt.Errorf("could not get slot for block %s: %w", parentHeader.Hash(), err)
+	}
+
+	if slotNumOfTheBestBlock == slotNum {
```

Comment:
should we check if `slotNumOfTheBestBlock >= slotNum` as well since it is possible to happen to build a block on slot `S` and the best block number was built in slot `S + 2`, right?

Reply:
> it is possible to happen to build a block on slot S and the best block number was built in slot S + 2

That does not seem possible. For that best block would have to be built in future, which does not make sense.

Reply:
Slot number are calculated using absolute clock and not logical clock. Each belongs to an actual time interval in the real world. So, if we are at slot S and out best block was built in slot S+2, that could mean one of three things
1) Best block is invalid. This would mean that we are not validating blocks properly. This would mean a bug in our code.
2) Our local time is not accurate.
3) We are not calculating slot number properly. This would mean a bug in our code.

If `slotNumOfTheBestBlock > slotNum` happens, I don't see if it makes sense to produce further blocks. It seems like the right place to panic.

Reply:
added a check for `slotNumOfTheBestBlock > slotNum`